### PR TITLE
Add Recursive Variable Extraction and Substitution

### PR DESCRIPTION
**Pull request by arnavs**
e.g. 

```
julia> expr
(((1 / β - 1) + δ) / α) ^ (1 / (α - 1))

julia> get_variables(expr)
3-element Array{Any,1}:
 β
 δ
 α
```

and 

```
julia> s = δ => γ
δ => γ

julia> ModelingToolkit.substitute_expr!(expr, s)

julia> expr
(((1 / β - 1) + γ) / α) ^ (1 / (α - 1))
```

### CODE REVIEW

**Comment by arnavs** [comment]
Also added a substitution thing.

File: src/utils.jl [hunk 1]
```diff
@@ -144,3 +144,26 @@ function vars!(vars, O)
 
     return vars
 end
+
+# variable extraction
+is_singleton(e) = hasproperty(e, :name) || hasproperty(e, :op) && hasproperty(e.op, :name)
+
+function get_variables(e::Expression, vars = nothing)
+  vars = isnothing(vars) ? [] : vars
```
Comment (ChrisRackauckas) [inline src/utils.jl]:
```suggestion
  vars = isnothing(vars) ? Variable[] : vars
```

Comment (ChrisRackauckas) [reply src/utils.jl]: Or do you want to extract the Operation of the variable? That may be more useful?

File: src/utils.jl [hunk 2]
```diff
@@ -144,3 +144,26 @@ function vars!(vars, O)
 
     return vars
 end
+
+# variable extraction
+is_singleton(e) = hasproperty(e, :name) || hasproperty(e, :op) && hasproperty(e.op, :name)
```
Comment (ChrisRackauckas) [inline src/utils.jl]:
```suggestion
is_singleton(e::Operation) = e.op isa Variable
```

File: src/utils.jl (hunk 1 above)
Comment (ChrisRackauckas) [inline src/utils.jl]:
```suggestion
  vars = isnothing(vars) ? Operation[] : vars
```

File: src/utils.jl (hunk 2 above)
Comment (ChrisRackauckas) [inline src/utils.jl]: then add `is_singleton(e) = false`

File: src/utils.jl [hunk 3]
```diff
@@ -144,3 +144,26 @@ function vars!(vars, O)
...
+  if is_singleton(e)
+    push!(vars, e)
+  elseif hasproperty(e, :args) # because it can sometimes be 0
+    foreach(x -> get_variables(x, vars), e.args)
+  end
+  return unique(vars)
+end
+
+# variable substitution
+function substitute_expr!(expr::Expression, s::Pair{Operation, Operation})
```
Comment (ChrisRackauckas) [inline src/utils.jl]: make a dispatch on Operation and then a dispatch on etc. and that would make this be a lot safer.

File: src/utils.jl (hunk 2 above)
Comment (arnavs) [reply src/utils.jl]: @ChrisRackauckas The issue with this is that `Variable`s don't have `Op`s, right?

```
julia> x = Variable(:x, known = true)
x (callable known variable)

julia> dump(x)
x (function of type Variable)
  name: Symbol x
  known: Bool true
```

Comment (arnavs) [reply src/utils.jl]: oh, I see what you mean. I wanted to leave it open to the possibility of non-Operation variables, but this works too.

Comment (arnavs) [reply src/utils.jl]: Done.

File: src/utils.jl (hunk 3 above)
Comment (arnavs) [reply src/utils.jl]: Done.

File: src/utils.jl [hunk 4]
```diff
@@ -144,3 +144,30 @@ function vars!(vars, O)
 
     return vars
 end
+
+# variable extraction
+is_singleton(e) = false
+is_singleton(e::Operation) = e.op isa Variable
+
+function get_variables(e::Expression, vars = nothing)
+    vars = isnothing(vars) ? Operation[] : vars
```
Comment (ChrisRackauckas) [inline src/utils.jl]:
```suggestion
function get_variables(e::Expression, vars = Operation[])
```

File: src/utils.jl [hunk 5]
```diff
@@ -144,3 +144,30 @@ function vars!(vars, O)
...
+    if e isa ModelingToolkit.Constant
```
Comment (ChrisRackauckas) [inline src/utils.jl]: different dispatch

Comment (arnavs) [reply src/utils.jl]: Done.

**Comment by ChrisRackauckas** [comment]
There was too much to do in this PR so I just finished it. Superseded by the squash merge https://github.com/JuliaDiffEq/ModelingToolkit.jl/commit/8c274a76ca027d83df6a9196a58416093a872ab5 . See if that works for you.

**Comment by arnavs** [comment]
Looks good to me.

**Pull request closed without merging**